samza-yarn/src/main/java/org/apache/samza/util/hadoop/CertFSFileSystem.java

```diff
@@ -0,0 +1,461 @@
+    }
+
+    try {
+      KeyPairGenerator keyGen = KeyPairGenerator.getInstance(CertFSConstants.KEY_ALGORITHM);//job config
+      keyGen.initialize(CertFSConstants.KEY_SIZE);
+      _keyPair = keyGen.generateKeyPair();
+    } catch (NoSuchAlgorithmException e) {
+      throw new IOException("Key generation cannot get the algorithm for " + CertFSConstants.KEY_ALGORITHM, e);
+    }
+
+    _uri = uri;
+
+    _workingDir = getHomeDirectory();
+  }
+
+  /**
+   * Return the protocol scheme for the FileSystem.
+   *
+   * @return <code>certfs</code>
+   */
+  @Override
+  public String getScheme() {
+    return _uri.getScheme();
+  }
+
+  @Override
+  public URI getUri() {
+    return _uri;
+  }
+
+  /**
+   * convert certfs URI to https URI
+   * @param certFsUri certFS URI where the CSR should be sent to
+   * @param excludesQuery whether or not to exclude the query params
+   * @return https URI
+   * @throws CertFSException when cannot construct https URI
+   */
+  private URI createHttpsUri(final URI certFsUri, final boolean excludesQuery) throws CertFSException{
+    String certFsScheme = certFsUri.getScheme();
+    if (!CertFSConstants.CERTFS_URI_SCHEME.equalsIgnoreCase(certFsScheme)) {
+      throw new CertFSException("Invalid scheme for certFsUri: " + certFsUri);
+    }
+    try {
+      String query = excludesQuery ? null : certFsUri.getQuery();
+      URI httpsUri = new URI(
+          CertFSConstants.HTTPS_URI_SCHEME,
+          certFsUri.getUserInfo(),
+          certFsUri.getHost(),
+          certFsUri.getPort(),
+          certFsUri.getPath(),
+          query,
+          certFsUri.getFragment()
+      );
+      return httpsUri;
+
+    } catch (URISyntaxException e) {
+      throw new CertFSException("cannot create the corresponding https URI due to syntax error for " + certFsUri, e);
+    }
+  }
+
+  /**
+   * get bcstyle x500name from config and replace the placeholder (such as __HOSTNAME__) if needed
+   * @return bcstyle X500Name in the format of string, e.g. "CN=myhost.com, OU=Security, O=LinkedIn, L=Mountain View, ST=California, C=US"
+   * @throws CertFSException when no valid bcstyle X500Name
+   * @throws IOException when generating hostname for bcstyle x500Name
+   */
+  private String getBcstyleX500Name() throws CertFSException, IOException {
+    // sample of raw bcstyle x500name
+    // "CN=__HOSTNAME__, OU=Security, O=LinkedIn, L=Mountain View, ST=California, C=US"
+    String bcstyleX500NameRaw = getConf().get(CertFSConstants.BCSTYLE_X500NAME);
+    if (null == bcstyleX500NameRaw) {
+      throw new CertFSException("no csr.ssl.bcstyle.x500name configured");
+    }
+    String bcstyleX500Name;
+    if (bcstyleX500NameRaw.indexOf(CertFSConstants.HOSTNAME_PLACEHOLDER) >=0) {//contains placeholder
+      bcstyleX500Name = bcstyleX500NameRaw.replace(CertFSConstants.HOSTNAME_PLACEHOLDER, getHostName());
+    } else {
+      bcstyleX500Name = bcstyleX500NameRaw;
```

Comment:
No, it has the __HOSTNAME__ placeholder and the code will generate hostname and replace the placeholder. But I am considering to move this out and leave it up to the user to implement this detail.